server/service/src/stock_line/update.rs

```diff
@@ -0,0 +1,108 @@
+use chrono::NaiveDate;
+use repository::{
+    RepositoryError, StockLine, StockLineRow, StockLineRowRepository, StorageConnection,
+};
+
+use crate::{
+    service_provider::ServiceContext,
+    stock_line::validate::{
+        check_location_exists, check_stock_line_exists, check_stock_line_on_hold, check_store,
+    },
+    SingleRecordError,
+};
+
+use super::query::get_stock_line;
+
+#[derive(Default, Debug, Clone, PartialEq)]
+pub struct UpdateStockLine {
+    pub id: String,
+    pub location_id: Option<String>,
+    pub cost_price_per_pack: Option<f64>,
+    pub sell_price_per_pack: Option<f64>,
+    pub expiry_date: Option<NaiveDate>,
+    pub on_hold: Option<bool>,
+    pub batch: Option<String>,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum UpdateStockLineError {
+    DatabaseError(RepositoryError),
+    StockDoesNotBelongToStore,
+    StockIsOnHold,
+    StockDoesNotExist,
+    LocationDoesNotExist,
+    UpdatedStockNotFound,
+}
+
+pub fn update_stock_line(
+    ctx: &ServiceContext,
+    input: UpdateStockLine,
+) -> Result<StockLine, UpdateStockLineError> {
+    use UpdateStockLineError::*;
+
+    let result = ctx
+        .connection
+        .transaction_sync(|connection| {
+            let existing = validate(connection, &ctx.store_id, &input)?;
+            let new_stock_line = generate(existing, input);
+            StockLineRowRepository::new(&connection).upsert_one(&new_stock_line)?;
+
+            get_stock_line(ctx, new_stock_line.id).map_err(|error| match error {
+                SingleRecordError::DatabaseError(error) => DatabaseError(error),
+                SingleRecordError::NotFound(_) => UpdatedStockNotFound,
+            })
+        })
+        .map_err(|error| error.to_inner_error())?;
+    Ok(result)
+}
+
+fn validate(
+    connection: &StorageConnection,
+    store_id: &str,
+    input: &UpdateStockLine,
+) -> Result<StockLineRow, UpdateStockLineError> {
+    use UpdateStockLineError::*;
+
+    let stock_line = check_stock_line_exists(connection, &input.id)?.ok_or(StockDoesNotExist)?;
+
+    if !check_store(&stock_line, store_id) {
+        return Err(StockDoesNotBelongToStore);
+    };
+    if check_stock_line_on_hold(&stock_line) {
+        return Err(StockIsOnHold);
+    }
+    if let Some(location_id) = input.location_id.clone() {
+        if !check_location_exists(connection, &location_id)? {
+            return Err(LocationDoesNotExist);
+        }
+    }
+
+    Ok(stock_line)
+}
+
+fn generate(
+    mut existing: StockLineRow,
+    UpdateStockLine {
+        id: _,
+        location_id,
+        cost_price_per_pack,
+        sell_price_per_pack,
+        expiry_date,
+        batch,
+        on_hold,
+    }: UpdateStockLine,
+) -> StockLineRow {
+    existing.location_id = location_id.or(existing.location_id);
```

Comment:
could you also add in logging please? Maybe that should be in a new PR though?

There is this sort of thing in the mSupply implementation:
```
	// 130328 v3.12 Sudhir #5209 : Log create, added source table and record id 
	If (Old([item_line]cost_price)#[item_line]cost_price)
		RELATE ONE([item_line]item_ID)
		$s_event:=("Modify cost price: "+[item]code+". Old "+String(Old([item_line]cost_price))+" New: "+String([item_line]cost_price))
		log_create("price change"; $s_event; [item_line]ID; Table(->[item_line]))
	End if 
	
	If (Old([item_line]sell_price)#[item_line]sell_price)
		RELATE ONE([item_line]item_ID)
		obStockLineForm.colombia.stockItemID:=[item]ID
		$s_event:=("Modify sell price: "+[item]code+". Old "+String(Old([item_line]sell_price))+" New: "+String([item_line]sell_price))
		log_create("price change"; $s_event; [item_line]ID; Table(->[item_line]))
	End if 
	
	If (Old([item_line]quantity)#[item_line]quantity)
		RELATE ONE([item_line]item_ID)
		$s_event:=("Modify quantity: "+[item]code+". Old "+String(Old([item_line]quantity))+" New: "+String([item_line]quantity))
		log_create("quantity change"; $s_event; [item_line]ID; Table(->[item_line]))
	End if 
	
	If (Old([item_line]available)#[item_line]available)
		RELATE ONE([item_line]item_ID)
		$s_event:=("Modify available: "+[item]code+". Old "+String(Old([item_line]available))+" New: "+String([item_line]available))
		log_create("quantity change"; $s_event; [item_line]ID; Table(->[item_line]))
	End if 
	
	If (Old([item_line]batch)#[item_line]batch)
		RELATE ONE([item_line]item_ID)
		$s_event:=("Modify batch: "+[item]code+". Old "+Old([item_line]batch)+" New: "+[item_line]batch)
		log_create("batch change"; $s_event; [item_line]ID; Table(->[item_line]))
	End if 
```

and

```
		log_create("Batch off hold"; "Item line #"+[item_line]ID+" Item "+[item]code; [item_line]ID; ```
```
		log_create("Batch set to hold"; "Item line #"+[item_line]ID+" Item "+[item]code; [item_line]ID; Table(->[item_line]))  // Always log putting a stock line on hold
```

Reply:
Do you want all the individual field changes logged? Or just have one log stating that the stock line has been updated? 🤔

Reply:
individual field changes.. if a field has changed. It's a bit clumsier to code and makes more log entries.. but I believe that mSupply has done it that way for a reason and the changing of stock line information is a bit sensitive, so would prefer to be cautious